### PR TITLE
`BlockId` removal: `BlockBuilderProvider::new_block_at`

### PR DESCRIPTION
**Pull request by michalkucharczyk**
It changes the arguments of `BlockBuilderProvider::new_block_at` from:
`BlockId<Block>` to: `Block::Hash`

This PR is part of BlockId::Number refactoring analysis (https://github.com/paritytech/substrate/issues/11292)

polkadot companion: paritytech/polkadot#6734
cumulus companion: paritytech/cumulus#2219

### CODE REVIEW

File: client/service/test/src/client/mod.rs [hunk 1]
```diff
@@ -1443,32 +1441,51 @@ fn respects_block_rules() {
...
 			.unwrap()
 			.build()
 			.unwrap()
 			.block;
+		let block_ok_1_hash = block_ok.hash();
 
 		let params = BlockCheckParams {
 			hash: block_ok.hash(),
-			number: 0,
+			number: 1,
```
Comment (michalkucharczyk) [inline client/service/test/src/client/mod.rs:1471]: Actually number given here does not matter, but the code is less confusing if the number actually matches the block.

File: client/service/test/src/client/mod.rs [hunk 2]
```diff
@@ -1483,34 +1500,35 @@ fn respects_block_rules() {
...
-			number: 1,
+			number: 2,
 			parent_hash: *block_ok.header().parent_hash(),
 			allow_missing_state: false,
 			allow_missing_parent: false,
 			import_existing: false,
 		};
 		if record_only {
-			fork_rules.push((1, block_ok.hash()));
+			fork_rules.push((2, block_ok.hash()));
```
Comment (michalkucharczyk) [inline client/service/test/src/client/mod.rs:1518]: Checking fork at block height 2 is more readable and less confusing, this is also inline with actual block number, and also with params passed to `check_block`.

**Review by bkchr — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2424115

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2424189

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2424190

**Review by skunert — APPROVED** [review]

**Comment by michalkucharczyk** [comment]
bot merge

**Pull request merged**